[Feature][Connector-V2][Timeplus] Initial source/sink connector for Timeplus

jovezhong opened the pull request:
### Purpose of this pull request
The initial working version for Timeplus sink with seatunnel framework. It closes https://github.com/apache/seatunnel/issues/7384


### Does this PR introduce _any_ user-facing change?
A new source and sink connector. The current testing and development are focusing on sink, but source should also work. README included.

### How was this patch tested?
End-to-end test to be added with testcontainer (waiting for testcontainer team to approve the PR)

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [x] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### Credits
This code of this PR is largely inspired by ClickHouse connector, StarRocks and Doris connector.

Commit:
fork connector-clickhouse

Commit:
rename

Commit:
rename

Commit:
move test code

Commit:
Update TimeplusFactoryTest.java

Commit:
Update TimeplusFactoryTest.java

Commit:
update pom.xml

Commit:
rename

Commit:
rename

Commit:
fix code style

Commit:
rename

Commit:
rename

Commit:
rename

Commit:
rename

Commit:
use proton java/jdbc

Commit:
rename

Commit:
remove unused jar

Commit:
fix the jdbc url

Commit:
Update error message

Commit:
rename the version properties in pom.xml

Commit:
Remove TimeplusFileSink

Commit:
rename

Commit:
rename

Commit:
WIP adding multi-table and create table support

Commit:
format code, passed compile

Commit:
WIP: set default values

Commit:
make TimeplusSinkFactory real work

Commit:
WIP: support datasavemode and schemasavemode

Commit:
add icecream for debug msg, WIP schema check

Commit:
WIP data type and SQL generating

Commit:
WIP: save mode

Commit:
about to remove shardMetadata

Commit:
WIP, pass values here and there

Commit:
working after fix the data type names in *InjectFunction

Commit:
format code and cleanup some logs

Commit:
No need to set primary key for auto-created stream

Commit:
fix the issue for data type lookup

Commit:
add debug message and keep renaming

Commit:
Update default value for bulk_size from 20K to 10K rows to avoid TOO_LARGE_RECORD error

Commit:
Squashed commit of the following:

commit 9d161d24ea2fed6ed7392deda6342fd068df3b6a
Author: lizhenglei <127465317+jackyyyyyssss@users.noreply.github.com>
Date:   Mon Aug 12 14:03:21 2024 +0800

    [Fix] Fix http e2e case (#7356)

    * 1

    * fix

    * fix

    * fix

    * fix

    * fix

    ---------

    Co-authored-by: lizhenglei <673421862@qq.com>

commit 2489f6446bae5b971a26e4cd3094f7d7af9d0208
Author: hailin0 <wanghailin@apache.org>
Date:   Mon Aug 12 13:36:15 2024 +0800

    [Improve][Connector] Add multi-table sink option check (#7360)

    * [Improve][Connector] Add multi-table sink option check

    * fix

commit fa34ac98b42cd2bf52837efa27382cfe24c655cd
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Mon Aug 12 13:35:43 2024 +0800

    [Improve][API] Check catalog table fields name legal before send to downstream (#7358)

    * [Improve][API] Check catalog table fields name legal before send to downstream

    * update

commit 068c5e3e3e8518e8cef2bd7599d85455d1adbaef
Author: Tyrantlucifer <TyrantLucifer@gmail.com>
Date:   Mon Aug 12 12:55:41 2024 +0800

    [Core][Flink] refactor flink proxy source/sink (#7355)

commit e23e3ac4ed5f83735797c4a9d5ffc0606cbe94ef
Author: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Date:   Sun Aug 11 17:40:02 2024 +0800

    Bump org.apache.activemq:activemq-client (#7323)

    Bumps org.apache.activemq:activemq-client from 5.14.5 to 5.15.16.

commit d46cf16e5a4bb0e4820feaf4dd03e7e9038d6281
Author: Jarvis <liunaijie1996@163.com>
Date:   Sat Aug 10 23:47:12 2024 +0800

    [Feature] Split transform and move jar into connectors directory (#7218)

commit 862e2055cec6a4428ec4cf67b472ba41ca364d05
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Sat Aug 10 15:21:36 2024 +0800

    [Improve][Document] Change deprecated connector name in setup.md (#7366)

commit 2769ed5029efc44891552725437e35d86a0eb93f
Author: hailin0 <wanghailin@apache.org>
Date:   Sat Aug 10 09:58:03 2024 +0800

    [Hotfix][Connector] Fix jdbc compile error (#7359)

commit 3130ae089e65bc71fa61784fad7d282f92aeadaf
Author: xxsc0529 <93303124+xxsc0529@users.noreply.github.com>
Date:   Fri Aug 9 20:02:06 2024 +0800

    [Fix][Connector-V2][OceanBase] Remove OceanBase catalog's dependency on mysql driver (#7311)

commit 921662722f01fd6e7bffb3b582e7624f21cddfee
Author: chaos <71205599+chaos-cn@users.noreply.github.com>
Date:   Fri Aug 9 13:53:28 2024 +0800

    Update ConsoleSinkFactory.java (#7350)

    Add existing rules to the optionRule

commit dc3c23981b3a78a3f1af69bcde57fbd5bf57d7b4
Author: dailai <dailai@chinatelecom.cn>
Date:   Fri Aug 9 13:11:30 2024 +0800

    [Improve][Jdbc] Skip all index when auto create table to improve performance of write (#7288)

commit 16950a67cd5990ed3510fd91564390e2ab2cd44c
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Thu Aug 8 17:27:47 2024 +0800

    [Improve][Connector-V2] Improve doris error msg (#7343)

commit a18fca80061925ba3a2c9bd152b450dcf3d9308b
Author: Alex Ting <alextinng@hotmail.com>
Date:   Thu Aug 8 11:22:41 2024 +0800

    [Fix][Connector-tdengine] Fix sql exception and concurrentmodifyexception when connect to taos and read data

commit 064fcad36d3f0d152b2ebc2be926ec0ed40972ab
Author: lizhenglei <127465317+jackyyyyyssss@users.noreply.github.com>
Date:   Wed Aug 7 21:57:57 2024 +0800

    [Improve][Transform] Improve DynamicCompile transform (#7319)

commit 16eeb1c12323b462474ad23103de96c0060a4537
Author: He Wang <wanghechn@qq.com>
Date:   Wed Aug 7 14:54:46 2024 +0800

    [Improve][Jdbc] Remove MysqlType references in JdbcDialect (#7333)

commit 191d9e18b911a6b17e804ea26da618ed6f3d2df5
Author: ChunFuWu <319355703@qq.com>
Date:   Wed Aug 7 14:53:51 2024 +0800

    [FIX][SFTP] Fix username parameter error in sftp sink document (#7334)

commit 764d8b0bc8a12100bf2ef07c9b1488ba2bda70ab
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Wed Aug 7 12:46:30 2024 +0800

    [Improve][API] Make sure the table name in TablePath not be null (#7252)

commit 891652399e8b97fb5cf1c7e6bdee87e7ec48469e
Author: hailin0 <wanghailin@apache.org>
Date:   Wed Aug 7 12:46:12 2024 +0800

    [Hotfix][Connector] Fix kafka consumer log next startup offset (#7312)

commit 819c6856513c9171b60d63c324e793a6eecbb9f6
Author: hailin0 <wanghailin@apache.org>
Date:   Wed Aug 7 12:45:54 2024 +0800

    [Improve][Jdbc] Merge user config primary key when create table (#7313)

commit 7c3cd99e007e2010fb88791cfd424d3642066a34
Author: hailin0 <wanghailin@apache.org>
Date:   Wed Aug 7 12:44:41 2024 +0800

    [Improve] Remove unused code (#7324)

commit 460e73ec36f1afac42c36320fb79a47a45bfc823
Author: hailin0 <wanghailin@apache.org>
Date:   Wed Aug 7 12:44:24 2024 +0800

    [E2E] Enable fakesource e2e of spark/flink (#7325)

commit f6a1e51b890582f1cf785a40b4bb38564c5652f8
Author: hailin0 <wanghailin@apache.org>
Date:   Wed Aug 7 12:44:13 2024 +0800

    [E2E] Enable JdbcPostgresIdentifierIT (#7326)

commit a12786b82101ac37859deeee521b59dfd6a9cd5f
Author: Carl-Zhou-CN <1058249259@qq.com>
Date:   Wed Aug 7 12:27:50 2024 +0800

    [DOC][Oss] fix document configuration is rectified when the oss is selected as the checkpoint base (#7332)

commit a4db64d7c76f07daaf8030011ee5e8552396acd2
Author: zhangdonghao <39961809+hawk9821@users.noreply.github.com>
Date:   Wed Aug 7 11:52:19 2024 +0800

    [Improve][E2E] Support windows for the e2e of paimon (#7329)

commit 855254e737051edbaf4ca08b95ca010fe18fd214
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Wed Aug 7 11:03:06 2024 +0800

    [Feature][Transform] Add LLM transform (#7303)

    * [Feature][Transform] Add LLM transform

    * update

    * update

    * retrigger

commit f0cefbeb4ac856f1eca17f068b2549d750f321ee
Author: Arin <136636751+asapekia@users.noreply.github.com>
Date:   Tue Aug 6 19:23:38 2024 +0530

    [Feature] [Activemq] Added activemq sink  (#7251)

commit bb2c912404fac13e829044c60259f2ab27bff3a1
Author: Guangdong Liu <804167098@qq.com>
Date:   Tue Aug 6 21:46:12 2024 +0800

    [Fix][Doris] Fix the abnormality of deleting data in CDC scenario. (#7315)

commit c94ea325b7a70ebbfc3eabdaa5025c6d0292384c
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Tue Aug 6 11:23:52 2024 +0800

    [Fix][Doc] Fix miss sink-options-placeholders.md in sidebars (#7310)

commit 73632bad2b93e6879c673e0f00bb83035aa51408
Author: hailin0 <wanghailin@apache.org>
Date:   Tue Aug 6 10:08:45 2024 +0800

    [Hotfix][Zeta] Fix task cannot be stopped when system is busy (#7292)

commit 9d56cc33b4f62316ed694c63a088fd50c1f51628
Author: virvle <55478661+virvle@users.noreply.github.com>
Date:   Mon Aug 5 17:51:26 2024 +0800

    Update Oracle-CDC.md (#7285)

commit 4f120ff34b1b5cf0536ff88cd9b571eff3b1fdcb
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Mon Aug 5 14:34:01 2024 +0800

    [Improve] Update snapshot version to 2.3.7 (#7305)

commit 2fd4eec22aeb3a35c558eeecbc35d35ad217db10
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Mon Aug 5 11:35:02 2024 +0800

    [Fix][Doc] Fix hybrid cluster deployment document display error (#7306)

commit 82f5d8c71212b0ceb4a57d9cc7cd97cab5aa6d4e
Author: hailin0 <wanghailin@apache.org>
Date:   Sat Aug 3 12:02:19 2024 +0800

    [Improve][SQL-Transform] Remove escape identifier from output fields (#7297)

commit 34a6b8e9f62d9e04da554062b46949c7075176ed
Author: Jast <admin@hadoop.wiki>
Date:   Wed Jul 31 14:43:11 2024 +0800

    [hotfix][connector-v2-hbase]fix and  optimize hbase source problem (#7148)

    * [hotfix][improve][doc]optimize connector hbase source

    * [doc]add dependent document

    * [doc]update dependent document

    * [improve]improve static use

    * [hotfix]add test case

    * [hotfix]add test case

    ---------

    Co-authored-by: Jia Fan <fanjiaeminem@qq.com>

commit 9df557cb12d39831a737658994e2fe2662230e1a
Author: lizhenglei <127465317+jackyyyyyssss@users.noreply.github.com>
Date:   Tue Jul 30 21:37:04 2024 +0800

    [Improve][Transform] Improve DynamicCompile transform  (#7264)

commit b9acb573b75055931ab5dc911aa05a04ea830c4f
Author: hailin0 <wanghailin@apache.org>
Date:   Tue Jul 30 16:12:57 2024 +0800

    [Feature][Transforms] Support cast to bytes function of sql (#7284)

commit ec1c3198bcffb8ff6b41014c36d4ddacedb03868
Author: Nothing. <oswinwuo.o@gmail.com>
Date:   Mon Jul 29 21:53:01 2024 +0800

    [Docs]translate event-listener doc into chinese (#7274)

commit f012b2a6f093bcad49fbd59946e75ce6d6a97838
Author: dailai <dailai@chinatelecom.cn>
Date:   Mon Jul 29 21:51:59 2024 +0800

    [Improve][Connector-v2] Optimize the way of databases and tables are checked for existence (#7261)

commit 48974917082ccc08f196e892697f4c518a8b0ebf
Author: 卢宗柱 <luzongzhu@sina.com>
Date:   Mon Jul 29 17:29:23 2024 +0800

    fix [Bug] Unable to create a source for identifier 'Iceberg'. #7182 (#7279)

    Co-authored-by: luzongzhu <luzongzhu@semi-tech.com>

commit 21c4f5245e8777a9b3a967b6038d00d3f20d9fb8
Author: corgy-w <73771213+corgy-w@users.noreply.github.com>
Date:   Mon Jul 29 17:28:10 2024 +0800

    [Improve][Core] Improve base on plugin name of lookup strategy (#7278)

    * [bug][plugin-discovery] fix multi plugin discovery

    * [bug][plugin-discovery] optimize code

    ---------

    Co-authored-by: wangchao <wangchao001@zhongan.com>

commit 3ccc6a8bd1e381a23e21eebcd9fca0d87694ba7c
Author: hailin0 <wanghailin@apache.org>
Date:   Mon Jul 29 10:41:09 2024 +0800

    [Hotfix][Zeta] Fix task cannot be stopped when system is busy (#7280)

commit d886495584c48392c2c09d2fb5d3d1780fc33244
Author: Guangdong Liu <804167098@qq.com>
Date:   Fri Jul 26 10:45:32 2024 +0800

    [Fix][Connector-V2] Fix doris TRANSFER_ENCODING header error (#7267)

commit 31e59cdf82e727a68aab57235eb99dec129cefea
Author: Xuzz <90698333+Xuzhengz@users.noreply.github.com>
Date:   Fri Jul 26 10:44:50 2024 +0800

    [Feature][Jdbc] Support hive compatibleMode add inceptor dialect (#7262)

commit d003bd85b65b307e77867781986cc7d58c3ba80a
Author: zhangdonghao <39961809+hawk9821@users.noreply.github.com>
Date:   Fri Jul 26 10:44:09 2024 +0800

    [Feature][Zeta] Added the metrics information of table statistics in multi-table mode (#7212)

commit 876d2f08d28313c963e520374b1d043f86d53302
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Thu Jul 25 10:24:55 2024 +0800

    [Fix][Dist] Remove aws jar dependency in release package (#7257)

    * [Fix][Dist] Remove aws jar dependency in release package

    * [Fix][Dist] Remove aws jar dependency in release package

commit 0d08b200619bdc2017236c46d4134e56f095e738
Author: dailai <dailai@chinatelecom.cn>
Date:   Thu Jul 25 10:00:13 2024 +0800

    [Improve][Connector-v2] Optimize the count table rows for jdbc-oracle and oracle-cdc (#7248)

commit cc5949988b7a851c49d594389fd3b96dbe0f9508
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Tue Jul 23 14:27:07 2024 +0800

    [Improve][Core] Move MultiTableSink to seatunnel-api module (#7243)

    * [Improve][Core] Move MultiTableSink to seatunnel-api module

    * [Improve][Core] Move MultiTableSink to seatunnel-api module

commit fe0c477b743e1d2b4c1dbc80b71e83b091306677
Author: Eric <gaojun2048@gmail.com>
Date:   Tue Jul 23 08:36:58 2024 +0800

    [Hotfix][Zeta] Fix taskgroup failed log lost (#7241)

commit b0fe432e9988693c7fa572782c815a04bbefdc05
Author: hawk9821 <39961809+hawk9821@users.noreply.github.com>
Date:   Mon Jul 22 16:12:12 2024 +0800

    The isNullable attribute is true when the primary key field in the Paimon table converts the Column object. #7231 (#7242)

commit 4b3af9bef4e1753838a7750e86bde71bae8562ae
Author: Guangdong Liu <804167098@qq.com>
Date:   Mon Jul 22 13:04:19 2024 +0800

    [Improve][Doris Connector] Unified serialization method,Use RowToJsonConverter and TextSerializationSchema (#7229)

    * 1

    * 1

    * 1

    * Update seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/source/serialization/RowBatch.java

    Co-authored-by: Jia Fan <fanjiaeminem@qq.com>

    ---------

    Co-authored-by: gdliu3 <gdliu3@iflytek.com>
    Co-authored-by: Jia Fan <fanjiaeminem@qq.com>

commit 4ec25f345f57debb1b48fcd96a723881161ff492
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Mon Jul 22 12:01:36 2024 +0800

    [Fix][Zeta] Fix release slot resource twice (#7236)

commit 7dc3fa8a13e77914798b133332a72831426cc935
Author: litiliu <38579068+litiliu@users.noreply.github.com>
Date:   Mon Jul 22 11:41:33 2024 +0800

    [Fix][Connector kafka]Fix Kafka consumer stop fetching after TM node restarted (#7233)

commit 063f83ca6cf97c0b574dda22c7908e81014cb165
Author: Guangdong Liu <804167098@qq.com>
Date:   Mon Jul 22 10:40:58 2024 +0800

    [Improve][Restapi] Add ip and port to monitoring information (#7203)

commit c4ca74122c21fd52cc98c7a2e604e24b0fa42911
Author: hailin0 <wanghailin@apache.org>
Date:   Sat Jul 20 21:00:06 2024 +0800

    [Feature][Core] Support using upstream table placeholders in sink options and auto replacement (#7131)

    、

commit 821dfc889c8a63f01ddfe3fd634384a0a1f12c88
Author: lizhenglei <127465317+jackyyyyyssss@users.noreply.github.com>
Date:   Sat Jul 20 20:25:50 2024 +0800

    [Feature][Transform] DynamicCompile add transform  (#7170)

commit 1e7c78dd53f6fd1e75806fc5f4fb3b26676d1f88
Author: hilo <1964736605@qq.com>
Date:   Fri Jul 19 19:44:18 2024 +0800

    [Improve][Doc] Add IMap OSS config jars into document (#7029)

commit d39ad93ad6c7420e42cd8965c9c3bf90f4a4e6b2
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Fri Jul 19 19:42:03 2024 +0800

    [Fix][Zeta] Fix hybrid deployment can not get worker when init (#7235)

commit 2058f5d9c31af97c39bfbad52e2e28cdaffa7892
Author: ChunFuWu <319355703@qq.com>
Date:   Fri Jul 19 14:21:23 2024 +0800

    [Chore][NOTICE] Correct year in NOTICE file (#7232)

commit 6e33a97c866d89c63793d297ab403e20c21d10d7
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Thu Jul 18 23:03:27 2024 +0800

    [Fix] Fix Hana type converter decimal scale is 0 convert to int error (#7167)

commit d1b4a7feb649f9e1095b6fe2e907202dd94b0896
Author: Jia Fan <fanjiaeminem@qq.com>
Date:   Thu Jul 18 15:31:30 2024 +0800

    [Improve][Zeta] Add check for submit duplicate job id (#7021)

commit e44e8b93bcc6c93806b1d25ea339f35d854c9897
Author: hailin0 <wanghailin@apache.org>
Date:   Thu Jul 18 15:30:49 2024 +0800

    [Improve][Jdbc] Support write unicode text into sqlserver (#7159)

commit f40f11af5d6c5f1ee29db2e59be59efa911efaaf
Author: tcodehuber <tcodehuber@gmail.com>
Date:   Thu Jul 18 00:03:33 2024 +0800

    [Docs] Update engine related docs info (#7228)

Commit:
add timeplus to plugin_config and plugin-mapping, also readme

Commit:
renames

Commit:
Remove port from TimeplusConfig

Commit:
format code

Commit:
format code

Commit:
fix import issue in TimeplusSink

Commit:
fix spotless warn, DO NOT USE import *

Commit:
Update Timeplus.md

Commit:
update release-note and dist/pom.xml

Commit:
Merge branch 'dev' into timeplus

Commit:
Merge branch 'dev' into timeplus

Commit:
remove an invalid dependency and add timeplus to label-scope-conf

Commit:
set default stream retention policy to 1GB 1HR

Commit:
Merge branch 'apache:dev' into timeplus

Commit:
Merge branch 'timeplus' of https://github.com/jovezhong/seatunnel into timeplus

Commit:
make the basic Timeplus source work

Commit:
upgrade timeplus http jdbc JAR version

Commit:
fix doc deadlink and refine source/sink readme

Commit:
Remove AutoService for TimeplusSink to fix CI check

Commit:
Merge branch 'dev' into timeplus

jovezhong commented:
Greetings, this PR was created 2 months with 2.3.6 and I have to keep rebasing it to support 2.3.7 and 2.3.8. In the meanwhile, I am waiting for a new release of testcontainers to run CI/UT more easily. https://github.com/testcontainers/testcontainers-java/commit/0f9956af805ceb08ae5f26dab0933cc8877afd54 was merged 3 weeks ago but no new release yet.

So I'd like to mark the PR ready for review. Please share your feedback and better to be included in 2.3.8.

Commit:
Update pom.xml

Hisoka-X commented:
> I am waiting for a new release of testcontainers to run CI/UT more easily. https://github.com/testcontainers/testcontainers-java/commit/0f9956af805ceb08ae5f26dab0933cc8877afd54 was merged 3 weeks ago but no new release yet

In SeaTunnel, many container used in test case not released in testcontainer project. We can just use `GenericContainer` with specific docker image name. For example, doris container not released in testcontainer. So we created it by `GenericContainer`. Please refer https://github.com/apache/seatunnel/blob/90cd46f50ac2f01e11ac5b7002688ea6c657cc82/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-doris-e2e/src/test/java/org/apache/seatunnel/e2e/connector/doris/AbstractDorisIT.java#L68

Hisoka-X commented:
> Please share your feedback and better to be included in 2.3.8.

I think we'll probably start releasing 2.3.8 by the end of September. So there's still time to add test cases.

Hisoka-X commented on docs/en/connector-v2/source/Timeplus.md:
```diff
@@ -0,0 +1,100 @@
+
+| Datasource | Supported Versions |                                                    Dependency                                                    |
+|------------|--------------------|------------------------------------------------------------------------------------------------------------------|
+| Timeplus | universal          | [Download](https://mvnrepository.com/artifact/org.apache.seatunnel/seatunnel-connectors-v2/connector-timeplus) |
```
```suggestion
| Timeplus | universal          | [Download](https://mvnrepository.com/artifact/org.apache.seatunnel/connector-timeplus) |
```

Hisoka-X commented on docs/en/connector-v2/sink/Timeplus.md:
```diff
@@ -0,0 +1,145 @@
+
+| Datasource | Supported Versions |                                                   Dependency                                                   |
+|------------|--------------------|----------------------------------------------------------------------------------------------------------------|
+| Timeplus   | universal          | [Download](https://mvnrepository.com/artifact/org.apache.seatunnel/seatunnel-connectors-v2/connector-timeplus) |
```
```suggestion
| Timeplus   | universal          | [Download](https://mvnrepository.com/artifact/org.apache.seatunnel/connector-timeplus) |
```

Hisoka-X reviewed (commented):
Does timeplus can directly write/read by clickhouse connector? I found many code come from clickhouse connector.

Commit:
Update docs/en/connector-v2/source/Timeplus.md

Co-authored-by: Jia Fan <fanjiaeminem@qq.com>

Commit:
Update docs/en/connector-v2/sink/Timeplus.md

Co-authored-by: Jia Fan <fanjiaeminem@qq.com>

jovezhong commented:
> Does timeplus can directly write/read by clickhouse connector? I found many code come from clickhouse connector.

Hi @Hisoka-X , yes, this timeplus connector is largely copied from the clickhouse connector and I added the multi-table sync and schema auto-created.  Timeplus Proton is built with Clickhouse libraries, but we cannot use the clickhouse JDBC driver or HTTP Java client to read/write Timeplus, because there are some significantly changes:
* the data type in ClickHouse are UInt8, we changed them to uint8
* function names in Timeplus are also consolidated to to_start_of_day() style
* also Timeplus supports streaming query and historical query. Running `select count() from sensor` will keep sending you the new count every X seconds. To run a bounded query, need to run `select count() from table(sensor)` or use the different HTTP/TCP port to turn on historical mode.

Hisoka-X commented:
> > Does timeplus can directly write/read by clickhouse connector? I found many code come from clickhouse connector.
> 
> Hi @Hisoka-X , yes, this timeplus connector is largely copied from the clickhouse connector and I added the multi-table sync and schema auto-created. Timeplus Proton is built with Clickhouse libraries, but we cannot use the clickhouse JDBC driver or HTTP Java client to read/write Timeplus, because there are some significantly changes:
> 
> * the data type in ClickHouse are UInt8, we changed them to uint8
> * function names in Timeplus are also consolidated to to_start_of_day() style
> * also Timeplus supports streaming query and historical query. Running `select count() from sensor` will keep sending you the new count every X seconds. To run a bounded query, need to run `select count() from table(sensor)` or use the different HTTP/TCP port to turn on historical mode.

Thanks @jovezhong for explain! Could you refactor code to reuse clickhouse connector code? Just like our file series connectors.

jovezhong commented:
Hi @Hisoka-X , I do want to reuse the code, but not sure about
1. how to manage the dependency, say the timeplus connector need to refer to clickhouse connector code, even some private/protected method/field
2. clickhouse connector doesn't support mutli-table sync or schema auto creation
3. there are many other details I need to customize the SQL or method, more than just JDBC driver class. For example, clickhouse connector uses lower level ClickHouse HTTP Client

Not sure whether there are tiers for new connectors. This one can be at technical preview, open for refinement, but we use this connector in some of our clients. Better to include in the built-in connector list.

Let me know whether it makes sense. Thanks for the review.

Hisoka-X commented:
> how to manage the dependency

We should have three modules.
```
connector-clickhouse-base (refer https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-file and https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-file/connector-file-base)

connector-clickhouse (dependency: connector-clickhouse-base refer https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-file/connector-file-local)
connector-timeplus (dependency: connector-clickhouse-base refer https://github.com/apache/seatunnel/tree/dev/seatunnel-connectors-v2/connector-file/connector-file-hadoop)
```

> even some private/protected method/field

We can change it to public if necessary, or create abstract class.

> clickhouse connector doesn't support mutli-table sync or schema auto creation

So I think we should put multi-table sync and schema auto creation part into `connector-timeplus`

> there are many other details I need to customize the SQL or method, more than just JDBC driver class. For example, clickhouse connector uses lower level ClickHouse HTTP Client

The core is not to make timeplus dependent on the clickhouse connector, but to hope that public methods can only exist in one copy for easy maintenance.

> Better to include in the built-in connector list.

Yes, after refactor, we can use clickhouse and timeplus at the same time.

jovezhong commented:
yes, I agree this will be a better solution. But I am not that familiar with seatunnel code base. Is that okay seatunnel core team to do the refactoring? I can provide testing env or support.

Hisoka-X commented:
> Is that okay seatunnel core team to do the refactoring?

Please can add test cases first so that we can ensure that there are no abnormalities in the function when we refactor. After this PR merge, we can create an issue to followup.

jovezhong commented:
Cool, I will add test cases in next 1 or 2 weeks (I will be on business trip next week)

Commit:
Avoid calling org.apache.logging.log4j.util.Strings.isBlank

May got java.lang.NoClassDefFoundError for Flink running

Commit:
fix issue for data type (found by e2e tests)

Commit:
add e2e test for Timeplus source/sink

Commit:
Update TimeplusIT.java

skip _tp_time and use table query

Commit:
format code

Commit:
Merge branch 'dev' into timeplus

jovezhong commented:
I am adding e2e test, almost done. But why I am getting the error for "On pull request update / Notify test workflow (pull_request_target) Failing after 8s"

> check_run_head: undefined
TypeError: Cannot read properties of undefined (reading 'head_sha')
    at eval (eval at callAsyncFunction (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:15143:16), <anonymous>:84:22)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async main (/home/runner/work/_actions/actions/github-script/v6/dist/index.js:15236:20)
Error: Unhandled error: TypeError: Cannot read properties of undefined (reading 'head_sha')

Hisoka-X commented:
![image](https://github.com/user-attachments/assets/838f6f53-7641-47cb-8517-b754bc57c078)
Looks like the github action startup failed.

Commit:
revert change for workflows/backend.yml

jovezhong commented:
I just reverted the change to the backend workflow. I am not sure whether this need to be changed in the PR or seatunnel team will update the github action. Let me check the CI build report.

jovezhong commented:
Got error

> Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.032 s <<< FAILURE! - in org.apache.seatunnel.api.connector.ConnectorSpecificationCheckTest
Error:  testAllConnectorImplementFactoryWithUpToDateMethod  Time elapsed: 0.469 s  <<< ERROR!
java.lang.ClassNotFoundException: org.apache.seatunnel.connectors.seatunnel.timeplus.sink.TimeplusSink

How to fix it?

I tested locally with `mvn -B -T 1 verify -DskipUT=true -DskipIT=false -D"license.skipAddThirdParty"=true -Dspotless.check.skip --no-snapshot-updates -pl :connector-timeplus-e2e -am -Pci`

Hisoka-X commented on pom.xml:
```diff
@@ -930,6 +930,8 @@
                     <java>
                         <excludes>
                             <exclude>src/main/java/org/apache/seatunnel/antlr4/generated/*.*</exclude>
+                            <exclude>**/TimeplusSinkFactory.java</exclude>
+                            <exclude>**/TimeplusSinkWriter.java</exclude>
```
Why?

jovezhong replied:
Good catch. I guess I added them during development to avoid checkstyle warning. I have removed these 2 lines

Hisoka-X commented:
> Got error
> 
> > Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.032 s <<< FAILURE! - in org.apache.seatunnel.api.connector.ConnectorSpecificationCheckTest
> > Error:  testAllConnectorImplementFactoryWithUpToDateMethod  Time elapsed: 0.469 s  <<< ERROR!
> > java.lang.ClassNotFoundException: org.apache.seatunnel.connectors.seatunnel.timeplus.sink.TimeplusSink
> 
> How to fix it?
> 
> I tested locally with `mvn -B -T 1 verify -DskipUT=true -DskipIT=false -D"license.skipAddThirdParty"=true -Dspotless.check.skip --no-snapshot-updates -pl :connector-timeplus-e2e -am -Pci`

Looks like you change `org.apache.seatunnel.connectors.seatunnel.timeplus.sink.TimeplusSink` to `org.apache.seatunnel.connectors.seatunnel.timeplus.sink.client.TimeplusSink`. Please clean your local cache.

Commit:
remove 2 excludes for style check

jovezhong commented:
Hi @Hisoka-X ,

> Looks like you change org.apache.seatunnel.connectors.seatunnel.timeplus.sink.TimeplusSink to org.apache.seatunnel.connectors.seatunnel.timeplus.sink.client.TimeplusSink. Please clean your local cache.

Since day 1, the classname is `org.apache.seatunnel.connectors.seatunnel.timeplus.sink.client.TimeplusSink`. May I know how `testAllConnectorImplementFactoryWithUpToDateMethod` determine the class name?

It seems that **SinkFactory has to be the same package as the **Sink

```java
                Class<? extends SeaTunnelSink> sinkClass =
                        (Class<? extends SeaTunnelSink>)
                                Class.forName(
                                        factory.getClass()
                                                .getName()
                                                .replace(
                                                        factoryName,
                                                        factoryName.replace("Factory", "")));
```

Hisoka-X commented:
> It seems that **SinkFactory has to be the same package as the **Sink

according the code, yes. cc @hailin0

jovezhong commented:
I can move TimeplusSinkFactory to the same package as TimeplusSink, or the other way. But this follows the same pattern as ClickhouseSink and ClickhouseSinkFactory. Is there similar issue for clickhouse sink?

Hisoka-X commented:
> I can move TimeplusSinkFactory to the same package as TimeplusSink, or the other way. But this follows the same pattern as ClickhouseSink and ClickhouseSinkFactory. Is there similar issue for clickhouse sink?

Clickhouse Sink not use https://github.com/apache/seatunnel/pull/7385/files#diff-72c6df20268df8e1150ccebc7bc1261adb2686de701e5b245d4789bcea79b911R103 at now. So we will not check it. We should update clickhouse sink in the future.

Commit:
move TimeplusSink from timeplus.sink.client to timeplus.sink

Otherwise testAllConnectorImplementFactoryWithUpToDateMethod will fail, since it assumes the factory and sink in the same package

Commit:
fix unit-test, add MULTI_TABLE_SINK_REPLICA to optional

Commit:
fix unit test, createWriter returns TimeplusSinkWriter not SinkWriter<..>

jovezhong commented:
Latest build error is from `SeaTunnel : E2E : Connector V2 : Amazon Dynamo DB .. FAILURE [03:35 min]` but I didn't change DynamoDB connector..

Hisoka-X commented:
> Latest build error is from `SeaTunnel : E2E : Connector V2 : Amazon Dynamo DB .. FAILURE [03:35 min]` but I didn't change DynamoDB connector..

Just re-trigger failed ci. It's unstable.

jovezhong commented:
Now I got some errors

> Test set: org.apache.seatunnel.connectors.seatunnel.timeplus.TimeplusIT
> -------------------------------------------------------------------------------
> Tests run: 7, Failures: 2, Errors: 1, Skipped: 0,
> org.awaitility.core.ConditionTimeoutException: 
> Assertion condition defined as a org.apache.seatunnel.e2e.common.container.seatunnel.SeaTunnelContainer There are still threads running in the container: 
> "ProtonClient4" #302 daemon prio=5 os_prio=0 tid=0x0000ffff1c76d800 nid=0x1a2 waiting on condition [0x0000ffff157fe000]
>    java.lang.Thread.State: WAITING (parking)
> 	at sun.misc.Unsafe.park(Native Method)
> 	- parking to wait for  <0x00000000f57445f0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)

> testTimeplus{TestContainer}[6]  Time elapsed: 21.398 s  <<< FAILURE!
> org.opentest4j.AssertionFailedError: id ==> expected: <1> but was: <0>

Some tests with Flink/Spark passed, some note. I guess there are some racing conditions or the ProtonClient is not closed. How to force close the running threads in SeaTunnelContainer?

Hisoka-X commented:
Is there no way to close the `ProtonClient4` thread at the end of the job/task?

jovezhong commented:
> Is there no way to close the `ProtonClient4` thread at the end of the job/task?

I checked all usage of ProtonClient in the connector code. They are either explicitly called with `close()` or put in ` try/catch` since it's AutoClosable.  Some of the end-to-end tests passed, some failed, not sure whether it's racing condition.

Hisoka-X commented:
>  Some of the end-to-end tests passed, some failed

Because we only test thread leak in zeta.

jovezhong commented:
I may not have time to look into this. Not sure whether whether we can merge the code as-is, or disable the e2e for flink/spark, and only keep it for zeta.

Hisoka-X commented:
I can take the time to follow up, but has to wait a few days. Adding in my TODO list.

jovezhong commented:
Hi @Hisoka-X , hope you had a great holiday. Is there anything I can do to merge the PR to catch up for the next release?

Hisoka-X commented:
> Hi @Hisoka-X , hope you had a great holiday. Is there anything I can do to merge the PR to catch up for the next release?

Sorry for not taking the time to follow up on this PR recently, I will refactor it tomorrow

jovezhong commented:
Hi @Hisoka-X , would you please review the PR. It has been over 3 months since I submitted the PR. The longer we wait, more conflicts the merge will be. Thanks and feel free to AT me if you have any questions/suggestions.

Hisoka-X committed:
Merge branch 'refs/heads/dev' into pr/7385

# Conflicts:
#	plugin-mapping.properties
